src/cache.rs

```diff
@@ -0,0 +1,225 @@
+// Copyright 2016 Mozilla
+//
+// Licensed under the Apache License, Version 2.0 (the "License"); you may not use
+// this file except in compliance with the License. You may obtain a copy of the
+// License at http://www.apache.org/licenses/LICENSE-2.0
+// Unless required by applicable law or agreed to in writing, software distributed
+// under the License is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR
+// CONDITIONS OF ANY KIND, either express or implied. See the License for the
+// specific language governing permissions and limitations under the License.
+
+use std::collections::BTreeMap;
+
+use rusqlite;
+
+use mentat_core::{
+    Entid,
+    TypedValue,
+};
+
+use mentat_db::cache::{
+    AttributeValueProvider,
+    Cacheable,
+    EagerCache,
+    CacheMap,
+};
+
+use errors::{
+    Result,
+};
+
+pub enum CacheAction {
+    Register,
+    Deregister,
+}
+
+#[derive(Clone)]
+pub struct AttributeCacher {
+    cache: BTreeMap<Entid, EagerCache<Entid, Vec<TypedValue>, AttributeValueProvider>>,   // values keyed by attribute
+}
+
+impl AttributeCacher {
+
+    pub fn new() -> Self {
+        AttributeCacher {
+            cache: BTreeMap::new(),
+        }
+    }
+
+    pub fn register_attribute<'sqlite>(&mut self, sqlite: &'sqlite rusqlite::Connection, attribute: Entid) -> Result<()> {
+        let value_provider = AttributeValueProvider{ attribute: attribute.clone() };
```

Comment:
Nit: space before `{`
I think `Entid` is `Copy`, so you don't need to write `clone()`.